### PR TITLE
check that we are not running multiple ranks on the same GPU

### LINKED ISSUE #190: Abort if multiple ranks are using the same GPU
cc @matthiasdiener 

Context: Lrun bug found by @matthiasdiener

### PR DESCRIPTION
**Pull request by matthiasdiener**
Fixes #190.

### CODE REVIEW

**Comment by matthiasdiener** [comment]
WIP - looking for feedback.

File: mirgecom/mpi.py [hunk 1]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
...
+            dev = cl_ctx.get_info(cl.context_info.DEVICES)[0]
+            platform = dev.get_info(cl.device_info.PLATFORM)
+            platform_name = platform.get_info(cl.platform_info.NAME)
+
+            if platform_name == "Portable Computing Language":
+                try:
+                    domain_id = dev.get_info(0x4010)
+                except cl._cl.LogicError:
+                    from warnings import warn
+                    warn("Can not detect whether multiple ranks are running on the"
```
Comment (inducer) [inline mirgecom/mpi.py:96]:
```suggestion
                    warn("Cannot detect whether multiple ranks are running on the"
```

File: mirgecom/mpi.py [hunk 2]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
...
+                         " same GPU because you need at least pocl version 1.7.")
```
Comment (inducer) [inline mirgecom/mpi.py:97]: As version advice goes, this is probably misleading: We'll likely patch this into our version of 1.6 on conda forge as soon as it's merged upstream. Maybe just link to the PR?

File: mirgecom/mpi.py [hunk 3]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
...
+        if size > 1:
+            node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+            node_rank = node_comm.Get_rank()
+            cl_ctx = cl.create_some_context()
+            dev = cl_ctx.get_info(cl.context_info.DEVICES)[0]
+            platform = dev.get_info(cl.device_info.PLATFORM)
+            platform_name = platform.get_info(cl.platform_info.NAME)
+
+            if platform_name == "Portable Computing Language":
+                try:
```
Comment (inducer) [inline mirgecom/mpi.py:92]: Probably no need to try this on devices that don't satisfy `bool(dev.type & cl.device_type.GPU)`.

File: mirgecom/mpi.py [hunk 4]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
                  "ranks. See https://mirgecom.readthedocs.io/en/latest/running.html#running-with-large-numbers-of-ranks-and-nodes"  # noqa: E501
                  " for more information.")
 
+        # Check whether multiple ranks are running on the same GPU on each node.
+        # Only works with pocl-cuda devices currently.
+        if size > 1:
+            node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+            node_rank = node_comm.Get_rank()
+            cl_ctx = cl.create_some_context()
+            dev = cl_ctx.get_info(cl.context_info.DEVICES)[0]
```
Comment (inducer) [inline mirgecom/mpi.py:87]: Should better handle multi-device contexts. (Maybe by just raising an error?)

File: mirgecom/mpi.py [hunk 5]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
...
+                    domain_id = dev.get_info(0x4010)
+                except cl._cl.LogicError:
+                    from warnings import warn
+                    warn("Can not detect whether multiple ranks are running on the"
+                         " same GPU because you need at least pocl version 1.7.")
+                else:
+                    bus_id = dev.get_info(cl.device_info.PCI_BUS_ID_NV)
+                    slot_id = dev.get_info(cl.device_info.PCI_SLOT_ID_NV)
+
+                    dev_id = hash(tuple((domain_id, bus_id, slot_id)))
```
Comment (inducer) [inline mirgecom/mpi.py:102]:
```suggestion
                    dev_id = (domain_id, bus_id, slot_id)
```

File: mirgecom/mpi.py [hunk 6]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
...
+
+                    if node_rank == 0:
+                        if len(dev_ids) != len(set(dev_ids)):
+                            print(dev_ids)
+                            raise RuntimeError(
+                                  "Multiple ranks are running on the same GPU.")
+
+            else:
+                from warnings import warn
+                warn("Can not detect whether multiple ranks are running on the "
```
Comment (inducer) [inline mirgecom/mpi.py:114]:
```suggestion
                warn("Cannot detect whether multiple ranks are running on the "
```

File: mirgecom/mpi.py [hunk 7]
```diff
@@ -78,6 +78,42 @@ def wrapped_func(*args, **kwargs):
...
+                     f" same GPU on platform '{platform_name}'.")
```
Comment (inducer) [inline mirgecom/mpi.py:115]: No point in issuing this warning unless we're talking to GPUs.

File: mirgecom/mpi.py (hunk 3 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: Fixed.

File: mirgecom/mpi.py (hunk 7 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: Fixed.

File: mirgecom/mpi.py (hunk 2 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: Fixed.

File: mirgecom/mpi.py (hunk 4 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: I implemented something, but not sure if that is what you had in mind.

File: mirgecom/mpi.py [hunk 8]
```diff
@@ -32,6 +32,66 @@
...
+            slot_id = dev.get_info(cl.device_info.PCI_SLOT_ID_NV)
+
+            dev_id = (domain_id, bus_id, slot_id)
+
+            dev_ids = node_comm.gather(dev_id, root=0)
+
+            if node_rank == 0:
+                if len(dev_ids) != len(set(dev_ids)):
+                    raise RuntimeError(
+                          "Multiple ranks are running on the same GPU.")
```
Comment (inducer) [inline mirgecom/mpi.py:87]: Print an example of which?

File: mirgecom/mpi.py [hunk 9]
```diff
@@ -32,6 +32,66 @@
...
+    dev = dev[0]
+    platform = dev.get_info(cl.device_info.PLATFORM)
+    platform_name = platform.get_info(cl.platform_info.NAME)
+
+    if not (dev.type & cl.device_type.GPU):
+        return
+
+    if platform_name == "Portable Computing Language":
+        try:
+            domain_id = dev.get_info(cl.device_info.PCI_DOMAIN_ID_NV)
```
Comment (inducer) [inline mirgecom/mpi.py:70]:
```suggestion
            domain_id = dev.pci_domain_id_nv
```
(same below)

File: mirgecom/mpi.py [hunk 10]
```diff
@@ -32,6 +32,66 @@
...
+        except cl._cl.LogicError:
+            from warnings import warn
+            warn("Cannot detect whether multiple ranks are running on the"
+                 " same GPU because you need a pocl version with"
+                 " https://github.com/pocl/pocl/pull/910.")
```
Comment (inducer) [inline mirgecom/mpi.py:75]: Give a version number indication.

File: mirgecom/mpi.py [hunk 11]
```diff
@@ -32,6 +32,66 @@
...
+    platform = dev.get_info(cl.device_info.PLATFORM)
+    platform_name = platform.get_info(cl.platform_info.NAME)
+
+    if not (dev.type & cl.device_type.GPU):
+        return
+
+    if platform_name == "Portable Computing Language":
+        try:
+            domain_id = dev.get_info(cl.device_info.PCI_DOMAIN_ID_NV)
+        except cl._cl.LogicError:
```
Comment (inducer) [inline mirgecom/mpi.py:71]: Wouldn't you get an `AttributeError` first, for old versions of pyopencl?

File: mirgecom/mpi.py [hunk 12]
```diff
@@ -32,6 +32,66 @@
...
+
+    node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+    node_rank = node_comm.Get_rank()
+
+    cl_ctx = cl.create_some_context()
+    dev = cl_ctx.get_info(cl.context_info.DEVICES)
+
+    # No support for multi-device contexts
+    if len(dev) > 1:
+        raise RuntimeError(f"Multiple devices selected for rank {rank}.")
```
Comment (inducer) [inline mirgecom/mpi.py:59]: Why? Shouldn't they be unique nonetheless?

File: mirgecom/mpi.py [hunk 13]
```diff
@@ -32,6 +32,66 @@
 import sys
 
 
+def _check_gpu_oversubscription():
+    """
+    Check whether multiple ranks are running on the same GPU on each node.
+
+    Only works with pocl-cuda devices currently due to the use of the
+    PCI_DOMAIN_ID_NV extension.
+    """
```
Comment (inducer) [inline mirgecom/mpi.py:41]: I could see maintaining this in PyOpenCL, too. What do you think of that?

File: mirgecom/mpi.py [hunk 14]
```diff
@@ -32,6 +32,66 @@
...
+    from mpi4py import MPI
+    import pyopencl as cl
+
+    size = MPI.COMM_WORLD.Get_size()
+    rank = MPI.COMM_WORLD.Get_rank()
+
+    if size <= 1:
+        return
+
+    node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
```
Comment (inducer) [inline mirgecom/mpi.py:51]: When does this communicator get destroyed?

**Review by inducer — COMMENTED** [review]
Thanks!

**Comment by matthiasdiener** [comment]
Tested on Lassen, ready for another review.

File: mirgecom/mpi.py (hunk 12 above)
Comment (inducer) [reply mirgecom/mpi.py:56]: Fair enough. I mean, handling multi-device contexts here isn't hard (just add all their IDs before gathering), but this is OK. If you do leave it, please change the error message to `NotImplementedError("multi-device contexts not yet supported")`.

File: mirgecom/mpi.py [hunk 15]
```diff
@@ -32,6 +32,64 @@
...
+        if node_rank == 0:
+            if len(dev_ids) != len(set(dev_ids)):
+                hostname = MPI.Get_processor_name()
+                dup = [item for item in dev_ids if dev_ids.count(item) > 1]
+
+                raise RuntimeError(
+                      f"Multiple ranks are sharing GPUs on node '{hostname}'."
+                      f" Duplicate PCIe IDs: {dup}.")
+
+    node_comm.Free()
```
Comment (inducer) [inline mirgecom/mpi.py:90]: This alloc/free pair should be a context manager.

File: mirgecom/mpi.py [hunk 16]
```diff
@@ -32,6 +32,64 @@
...
+    node_rank = node_comm.Get_rank()
+
+    try:
+        domain_id = hex(dev.pci_domain_id_nv)
+    except (cl._cl.LogicError, AttributeError):
+        from warnings import warn
+        warn("Cannot detect whether multiple ranks are running on the"
+             " same GPU because it requires Nvidia GPUs running with"
+             " pyopencl>2021.1.1 and (Nvidia CL or pocl>1.6).")
+    else:
```
Comment (inducer) [inline mirgecom/mpi.py:73]: Just return from the handler and avoid the `else`. (requires the context mgr to be correct)

File: mirgecom/mpi.py [hunk 17]
```diff
@@ -32,6 +32,64 @@
...
+    node_comm = MPI.COMM_WORLD.Split_type(MPI.COMM_TYPE_SHARED)
+    node_rank = node_comm.Get_rank()
+
+    try:
+        domain_id = hex(dev.pci_domain_id_nv)
+    except (cl._cl.LogicError, AttributeError):
+        from warnings import warn
+        warn("Cannot detect whether multiple ranks are running on the"
+             " same GPU because it requires Nvidia GPUs running with"
+             " pyopencl>2021.1.1 and (Nvidia CL or pocl>1.6).")
```
Comment (inducer) [inline mirgecom/mpi.py:72]: It will say this for AMD and Intel GPUs now. Is that intended?

Comment (matthiasdiener) [reply mirgecom/mpi.py:72]: Do you mean this warning should not be shown for AMD/Intel GPUs? Or is the text unclear?

File: mirgecom/mpi.py (hunk 12 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: Changed the error message.

File: mirgecom/mpi.py (hunk 13 above)
Comment (inducer) [reply mirgecom/mpi.py]: OK. Nvm for now.

File: mirgecom/mpi.py (hunk 15 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: I didn't find much info regarding mpi4py context managers so I implemented a simple one.

File: mirgecom/mpi.py (hunk 16 above)
Comment (matthiasdiener) [reply mirgecom/mpi.py]: Fixed.

**Comment by matthiasdiener** [comment]
Ready for another review

File: mirgecom/mpi.py (hunk 17 above)
Comment (inducer) [reply mirgecom/mpi.py]: Fair enough.

**Review by inducer — APPROVED** [review]

**Comment by inducer** [comment]
Thanks!

**Pull request merged**